core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/UnitOfWork.java

```diff
@@ -0,0 +1,156 @@
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.user.core.util;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import javax.sql.DataSource;
+import java.sql.Connection;
+import java.sql.SQLException;
+
+/**
+ * Support class to implement Unit of work patter.
+ */
+public class UnitOfWork {
+
+    private static Log log = LogFactory.getLog(UnitOfWork.class);
+    private boolean errorOccurred = false;
+    private static ThreadLocal<UnitOfWorkTransactionContext> transactionContextThreadLocal = new ThreadLocal<>();
+
+    public UnitOfWork() {
+
+        super();
+    }
+
+    /**
+     * Begin the transaction process.
+     */
+    public static UnitOfWork beginTransaction() {
+
+        UnitOfWorkTransactionContext UnitOfWorkTransactionContext = transactionContextThreadLocal.get();
+        if (UnitOfWorkTransactionContext == null) {
+            UnitOfWorkTransactionContext = new UnitOfWorkTransactionContext();
+            transactionContextThreadLocal.set(UnitOfWorkTransactionContext);
+        }
+        UnitOfWorkTransactionContext.incrementTransactionDepth();
+        return new UnitOfWork();
+    }
+
+    /**
+     * Returns an database connection.
+     *
+     * @param dataSource dataSource of the connection.
+     * @return current connection
+     * @throws SQLException
+     * @Deprecated The getDBConnection should handle both transaction and non-transaction connection. Earlier it
+     * handle only the transactionConnection. Therefore this method was deprecated and changed as handle both
+     * transaction and non-transaction connection. getDBConnection(DataSource dataSource, boolean autoCommit) method
+     * used as alternative of this method.
+     */
+    @Deprecated
+    public Connection getDBConnection(DataSource dataSource) throws SQLException {
+
+        return getDBConnection(dataSource, true);
+    }
+
+    /**
+     * Returns an database connection.
+     *
+     * @param dataSource dataSource of the connection.
+     * @param autoCommit autocommit state of the connection.
+     * @return current connection
+     * @throws SQLException
+     */
+    public Connection getDBConnection(DataSource dataSource, boolean autoCommit) throws SQLException {
+
+        UnitOfWorkTransactionContext unitOfWorkTransactionContext = transactionContextThreadLocal.get();
+        if (unitOfWorkTransactionContext == null) {
+            throw new UnitOfWorkException("There is no transaction getting started");
```

Comment:
throw new UnitOfWorkException("Can not get a connection as Transaction context is not started for datasource : "+ dataSource );